resharper/resharper-unity/src/CSharp/Feature/Services/QuickFixes/MoveQuickFixes/MonoBehaviourMoveUtil.cs

```diff
@@ -0,0 +1,263 @@
+                if (declaredElement == null || isElementIgnored != null && isElementIgnored(declaredElement))
+                    continue;
+
+                // if declared element is local, we can't move our expression outside the method
+                switch (declaredElement)
+                {
+                    case ILocalVariableDeclaration _:
+                    case IParameterDeclaration _:
+                    case ILocalFunctionDeclaration _:
+                    case IDelegateDeclaration _:
+                    case ILambdaExpression _:
+                    case IAnonymousFunctionExpression _:
+                    case IQueryParameterPlatform _:
+                        return false;
+                }
+            }
+            return true;
+        }
+
+        public static void MoveToMethodWithFieldIntroduction([NotNull]IClassDeclaration classDeclaration, [NotNull]ICSharpExpression expression, [NotNull] string methodName, string fieldName = null)
+        {
+            classDeclaration.GetPsiServices().Locks.AssertReadAccessAllowed();
+            classDeclaration.GetPsiServices().Locks.AssertMainThread();
+            
+            var methodDeclaration = GetOrCreateMethod(classDeclaration, methodName);
+            MoveToMethodWithFieldIntroduction(classDeclaration, methodDeclaration, expression, fieldName);
+        }
+        
+        public static void MoveToMethodWithFieldIntroduction([NotNull]IClassDeclaration classDeclaration,[NotNull] IMethodDeclaration methodDeclaration,
+            [NotNull]ICSharpExpression expression, string fieldName = null)
+        {
+            classDeclaration.GetPsiServices().Locks.AssertReadAccessAllowed();
+            classDeclaration.GetPsiServices().Locks.AssertMainThread();
+            
+            var result = GetDeclaredElementFromParentDeclaration(expression);
+            
+            var factory = CSharpElementFactory.GetInstance(classDeclaration);
+
+            var type = expression.Type(new ResolveContext(classDeclaration.GetPsiModule()));
+            if (type.IsUnknown)
+                type = TypeFactory.CreateTypeByCLRName("System.Object", classDeclaration.GetPsiModule());
+
+            var baseName = fieldName ?? CreateBaseName(expression, result);
+            var name = NamingUtil.GetUniqueName(expression, baseName, NamedElementKinds.PrivateInstanceFields, de => !de.Equals(result));
+
+            var isVoid = type.IsVoid();
+
+            if (!isVoid)
+            {
+                var field = factory.CreateFieldDeclaration(type, name);
+                field.SetAccessRights(AccessRights.PRIVATE);
+
+                classDeclaration.AddClassMemberDeclaration(field);
+            }
+
+            var initialization = isVoid ? factory.CreateStatement("$1;", name, expression.Copy()) : 
+                factory.CreateStatement("$0 = $1;", name, expression.Copy());
+            
+            var body = methodDeclaration.EnsureStatementMemberBody();
```

Comment:
I wonder how this would/should work with code like this:

```cs
class MonoBehaviour
{
  public virtual extern void Start();
}

abstract class C : MonoBehaviour
{
  public abstract override void Start(); // move something to this method
}
```